### ISSUE TITLE
3DSecure will not show when using drop-in (and possible solution)

### ISSUE THREAD

**Issue by meismyles** [braintree/braintree-ios-drop-in#26]
Currently 3D Secure does not seem to show when using the iOS drop-in.

The documentation advises that only the following is required to make it work:
```swift
let request =  BTDropInRequest()
request.threeDSecureVerification = true
request.amount = "1.00"
```

Setting `request.threeDSecureVerification = true` does cause the 3D checks to be started when tapping 'Add Card' in the drop-in UI, however, the loading indicator spins and then nothing happens.

Looking at the SDK code, it appears that in `BTThreeDSecureDriver.m` within the actual Braintree Core library, the following function is called at line 94:
```objc
[self informDelegateRequestsPresentationOfViewController:navigationController];
```

The function that is called (and its sibling dismiss function) both use `respondsToSelector` checks to make sure that the delegate does actually implement the required `paymentDriver` functions that are used to show the 3DS dialogue:
```objc
#pragma mark Delegate informer helpers

- (void)informDelegateRequestsPresentationOfViewController:(UIViewController *)viewController {
    if ([self.delegate respondsToSelector:@selector(paymentDriver:requestsPresentationOfViewController:)]) {
        [self.delegate paymentDriver:self requestsPresentationOfViewController:viewController];
    }
}

- (void)informDelegateRequestsDismissalOfViewController:(UIViewController *)viewController {
    if ([self.delegate respondsToSelector:@selector(paymentDriver:requestsDismissalOfViewController:)]) {
        [self.delegate paymentDriver:self requestsDismissalOfViewController:viewController];
    }
}
```

At this stage when using the drop-in UI, the delegate in question is an instance of `BTCardFormViewController.m` (since this is currently showing the 'Add Card' dialogue to the user). This class does not implement those functions and so the check fails and 3DS is never shown.

It seems to me that this should implement these functions or 3DS is never going to work in the drop-in. [I have submitted a pull request the changes that I believe should fix it.](https://github.com/braintree/braintree-ios-drop-in/pull/25)

Please take a look and get back to me. Cheers! 👍

**Comment by meismyles** [comment]
Not sure if I am correct here still, but assuming I am, the above solution does indeed show 3DS, but creates a new issue in that the actual result of the 3DS dialogue (success, failure, cancelled) doesn't seem to be accessible from anywhere. Without this result, we are unable to determine where to take the user next.

**Comment by meismyles** [comment]
Following on from my previous comment regarding what happens with the result of the 3DS web view dialogue - when entering a password and tapping submit, or tapping cancel on the nav bar it seems that the following functions are used within `BTThreeDSecureDriver.m` starting line 173:

```objc
#pragma mark BTThreeDSecureAuthenticationViewControllerDelegate

- (void)threeDSecureViewController:(__unused BTThreeDSecureAuthenticationViewController *)viewController
               didAuthenticateCard:(BTThreeDSecureCardNonce *)tokenizedCard
                        completion:(void (^)(BTThreeDSecureViewControllerCompletionStatus))completionBlock
{
    self.upgradedTokenizedCard = tokenizedCard;
    completionBlock(BTThreeDSecureViewControllerCompletionStatusSuccess);
    [self.apiClient sendAnalyticsEvent:@"ios.threedsecure.authenticated"];
}

- (void)threeDSecureViewController:(__unused BTThreeDSecureAuthenticationViewController *)viewController
                  didFailWithError:(NSError *)error {
    if ([error.domain isEqualToString:BTThreeDSecureErrorDomain] && error.code == BTThreeDSecureErrorTypeFailedAuthentication) {
        [self.apiClient sendAnalyticsEvent:@"ios.threedsecure.error.auth-failure"];
    } else {
        [self.apiClient sendAnalyticsEvent:@"ios.threedsecure.error.unrecognized-error"];
    }

    self.upgradedTokenizedCard = nil;
    self.completionBlockAfterAuthenticating(nil, error);
}

- (void)threeDSecureViewControllerDidFinish:(BTThreeDSecureAuthenticationViewController *)viewController {
    if (self.upgradedTokenizedCard) {
        self.completionBlockAfterAuthenticating(self.upgradedTokenizedCard, nil);
    } else {
        self.completionBlockAfterAuthenticating(nil, nil);
        [self.apiClient sendAnalyticsEvent:@"ios.threedsecure.canceled"];
    }

    self.completionBlockAfterAuthenticating = nil;
    [self informDelegateRequestsDismissalOfViewController:viewController];
}

- (void)threeDSecureViewController:(__unused BTThreeDSecureAuthenticationViewController *)viewController
      didPresentErrorForURLRequest:(NSURLRequest *)request {
    [self.apiClient sendAnalyticsEvent:[NSString stringWithFormat:@"ios.threedsecure.error.webview-error.%@", request.URL.host]];
}
```

Entering the correct password calls the following as expected:
```objc
- (void)threeDSecureViewController ... didAuthenticateCard ... {
```

However, typing in an incorrect password still calls back the above function and not: 
```objc
- (void)threeDSecureViewController ... didFailWithError ... {
```
as would be expected.

Also, if pressing cancel, the following function is called:
```objc
- (void)threeDSecureViewController ... threeDSecureViewControllerDidFinish ... {
```
which I suppose could be expected behaviour. However, this function passes `nil` as the `error` parameter of the completion block:
```objc
...
self.completionBlockAfterAuthenticating(nil, nil);
...
```
which means when the completion block of `BTDropInController` is called, there is no way of us telling whether the user cancelled the 3DS check.

**Comment by meismyles** [comment]
Also, when entering the correct 3DS password, and the following is called:
```objc
- (void)threeDSecureViewController ... didAuthenticateCard ... {
```
It appears that the upgraded nonce that should now have 3DS attached to it, still matches the original non-3DS nonce exactly. I would have thought that the upgraded nonce should be different? Or should they be the same?

**Comment by meismyles** [comment]
FYI: All of the above is being carried out in Sandbox, using test cards referenced in the [Cardinal Centinel for Merchants test cases document](https://developers.braintreepayments.com/files/Centinel.IntegrationGuide.ConsumerAuthentication_TestCases_v1_18_0_20160823.pdf).

**Comment by meismyles** [comment]
Minor update to the above. I had said it was not possible to determine if the user had success with 3DS, or failure or cancelled.

We can actually determine if the user has cancelled, as the returned nonce in now nil.

However, we cannot determine whether there was success or failure without further modifications to the Braintree SDK (so that an error is returned along with the non-upgraded nonce).

**Comment by sdcoffey** [comment]
Hey @meismyles,

Thanks for bringing this up (and submitting a PR)! I'm taking a look at this right now.

**Comment by sdcoffey** [comment]
Hey @meismyles,

So I looked through your PR, tested it out and it looks good. I guess my question now is why you need to know if the password the user entered in the 3DS auth webview was correct or not. This information is (and should remain) opaque to us client-side. My sense it that a generic auth error is returned when the user enters an incorrect password. 

For example, if you use the PAN `4000 0000 0000 0028` with expiration `01/20`, Cardinal will simulate an authentication error in sandbox, and you'll see that error come back in the callback of `BTDropInController initWithAuthorization: request: handler:`. Unless I'm misunderstanding your use case, this should be sufficient to determine that authentication failed without providing more specific information about the password. Let me know if that makes sense, and again,  thanks for finding this and making a PR!

**Comment by meismyles** [comment]
Hey @sdcoffey,

Sweet ok!

Yeah, my bad, I think you are correct here. When testing with the test cards that simulate success, I thought that since the password was the note, this still needed to be correct. However, when entering any password (correct or incorrect) and pushing submit, the web view dismissed regardless.

When testing in a production environment with a real card, red text actually appears within the 3DS web page itself telling you the password was incorrect if you get it wrong and you can't seem to proceed. So I guess the following could happen:
* We will get the success callback with a valid nonce meaning 3DS was verified correctly.
* The user tapped cancel (on the navbar) and the nonce will be nil so we can detect that.
* The user tapped cancel within the 3DS web page itself - in which case I'm not sure what happens at the moment. Will need to check this tomorrow.
* Some kind of other non-password related error which will be caught by the `didFailWithError` function.

One other issue that has cropped up as a result of this is to do with card vaulting. So it appears that once the user taps 'Add Card', immediate communication is made with Braintree and at this stage the users card is vaulted if configured to do so. This occurs **before** the 3DS popup has been shown. If the user now fails to verify with through 3DS successfully, or if they decide to cancel, their card information has still been vaulted and is shown in recent/saved cards. This feels odd and it seems to me that cards should only be saved at the point the user feels they are now going to be charged. If I tapped cancel on 3DS, I would assume my current transaction was being cancelled and would thus not expect my card to have been saved. Any thoughts?

**Comment by meismyles** [comment]
Also, this fix likely to be pulled into the CocoaPods version soon? Or any timeframe?

**Comment by meismyles** [comment]
In fact, doesn't matter. Need to use my fork for now anyway as it has another slight modification in it.

**Comment by meismyles** [comment]
Further to the vaulting issue above. I'm thinking now that since 3DS probably doesn't show until after a nonce is returned by Braintree, it's not easy to change this.

However, since pretty much everywhere else in the documentation it specifies that vaulting is done server side (and in fact we do already perform vaulting server side), it seems kind of odd to have this in the drop-in anyway. It seems to me it would make more sense to not have this automatic vaulting when using the drop-in and just let us do it server side?

Also, I've not tested, but I'm assuming there wont be any issues with a kind of 'double vaulting' since the drop-in will attempt to vault (when we get the first nonce back from Braintree), and then our server will also be trying to vault it once the order has been completed successfully?

**Comment by sdcoffey** [comment]
Hey @meismyles,

I realize our docs are a little unclear on this, so let me give you a brief overview of how vaulting works. You are correct that merchants can vault cards server-side, and that this is probably the biggest use case, but we also chose to enable vaulting client-side in Drop-in [with the proper credientials](https://developers.braintreepayments.com/guides/authorization/overview#capabilities), to provide the best experience for the end user. You're also right about idempotency. If a user adds a card in Drop-in, which is the vaulted, and then you vault it _again_ server-side, there will still only be one instance of that card in the vault.

Re: 3DS in Drop-in, with your PR, we'll always verify the card when it's added the first time by the user. However, if you have 3DS enabled, you _always_ need to check if a liability shift is possible, and if so, walk the user through the 3DS verification a before you can transact with that nonce. I realize our docs might be a little unclear on this, but [here's](https://developers.braintreepayments.com/guides/3d-secure/client-side/ios/v4#advanced-client-side-options) how you can check for a liability shift, and use `BTThreeDSecureDriver` to perform additional verifications.

Let me know if this is unclear, or if you have any ideas on how we can improve our documentation, we really appreciate the thoroughness here, and the willingness to help improve the product!

**Comment by demerino** [comment]
>  you always need to check if a liability shift is possible

Adding to what @sdcoffey said - you always need to run the `threeDSecureVerification` on a nonce from Drop-in and check the result to see if further verification is necessary.

**Comment by meismyles** [comment]
@sdcoffey @demerino Ok cheers guys, got a boat load of questions for you here as it's all getting confusing hah. Apologies about the length of this!

What seems very unclear to me still is if using the drop-in, what parts of [this document](https://developers.braintreepayments.com/guides/3d-secure/client-side/ios/v4) are relevant. At the start of the doc, it says:

>You can structure your native checkout flow however you'd like. If you don't already have a credit card form, consider our Drop-in UI.

This statement seems to imply that the rest of the page is not relevant if you use the drop-in. So maybe you could re-word the doc here to make it more clear.

Following that, it shows the code examples under the following titles [Render a checkout page](https://developers.braintreepayments.com/guides/3d-secure/client-side/ios/v4#render-a-checkout-page), [Verify a credit card](https://developers.braintreepayments.com/guides/3d-secure/client-side/ios/v4#verify-a-credit-card) and [Advanced client-side options](https://developers.braintreepayments.com/guides/3d-secure/client-side/ios/v4#advanced-client-side-options).

1. So, starting with [Render a checkout page](https://developers.braintreepayments.com/guides/3d-secure/client-side/ios/v4#render-a-checkout-page), the code in here initialises a `BTThreeDSecureDriver`. Since an instance of this class is initialised at some stage by the Drop-In, this section of the documentation does not seem relevant if using the Drop-In. Is this correct?
2. [Verify a credit card](https://developers.braintreepayments.com/guides/3d-secure/client-side/ios/v4#verify-a-credit-card) - so the first code example in this section initialises a `BTCard` object which then has its `number`, `expirationMonth` and `expirationYear` properties set based on the _'UI State'_. The `tokenizeCard` function is then called, passing in this `BTCard` object.
This all again seems to not be relevant if using the drop in because **(a)** we don't have access to the UI state/card information being entered as the Drop-In is handling it, and **(b)** the `tokenizeCard` function is called at some stage during the Drop-In flow anyway. Is this correct?
3. The next code block has the 2 `paymentDriver` functions which were implemented in [my pull request yesterday](https://github.com/braintree/braintree-ios-drop-in/pull/25).
So these again do not seem relevant as the Drop-In is now presenting the 3D Secure dialogue (`BTThreeDSecureAuthenticationViewController`) over the 'Add Card' view controller (`BTCardFormViewController`) rather than the developer doing it manually. Correct?
4. The next code example shows the `verifyCard` function being called. Again, this does not seem relevant if using the Drop-In since it is called at some stage during the flow if `request.threeDSecureVerification = true` was set and used to initialise the `BTDropInController`. Is this correct?

Now the following section - [Advanced client-side options](https://developers.braintreepayments.com/guides/3d-secure/client-side/ios/v4#advanced-client-side-options) - does seem necessary whether using the Drop-In or not based on your comments above (but I'm not sure it is possible).

5. The code example shows the following:
```swift
// In this example, `card` is a `BTThreeDSecureCardNonce`.
if card.liabilityShiftPossible && card.liabilityShifted {
    print("liability shift possible and liability shifted")
} else {
    print("3D Secure authentication was attempted but liability shift is not possible")
}
```
So, the card object here `BTThreeDSecureCardNonce` is a subclass of `BTCardNonce` which in turn is a subclass of `BTPaymentMethodNonce`. The 2 `BOOL`s, `liabilityShifted` and `liabilityShiftPossible` are properties of the `BTThreeDSecureCardNonce` object. However, by the time:
```objc
BTDropInController.m - line 268

- (void)cardTokenizationCompleted:(BTPaymentMethodNonce *)tokenizedCard error:(NSError *)error sender:(BTCardFormViewController *)sender {
```
is called this `BTThreeDSecureCardNonce` has been cast all the way up to its parent parent type `BTPaymentMethodNonce`. This is then returned in the completion to where we (the developer) first initialised our `BTDropInController` in the completion block `(controller, result, error)` - specifically, it is accessible through `result.paymentMethod`. This is the first time we (the developer) have access to any information since initialising the Drop-In - and since we can only see the `BTPaymentMethodNonce` object, we don't have access to the 2 `liabilityShift...` check `BOOL`s meaning it's not possible for us to do the check.

It seems at this stage, I guess we could potentially try casting the `BTPaymentMethodNonce` object back down to a `BTThreeDSecureCardNonce`, and if it works, we can access those `BOOL`s and proceed knowing 3DS has been shown. Or, if the cast fails, then it never was a `BTThreeDSecureCardNonce` in the first place, so we know 3DS didn't show. However, this feels like a bit of a workaround and seems as though it would make more sense for the SDK to be modified so that we (the developer) can somehow access the `BTThreeDSecureCardNonce` object without any casting if 3DS was carried out. So, as it stands, it would appear it's not easily possible to do the liability shift checks, would you agree? And any further thoughts?

6. @sdcoffey, you say above:
>However, if you have 3DS enabled, you always need to check if a liability shift is possible, and if so, walk the user through the 3DS verification a before you can transact with that nonce. I realise our docs might be a little unclear on this, but here's how you can check for a liability shift, and use BTThreeDSecureDriver to perform additional verifications.

So, apologies if I'm not understanding you correctly here, but you seem to say that we first check if a liability shift is possible, and if so, we then display the 3DS dialogue to the customer. However, based on how the SDK seems to work, we only know this information **after** the 3DS dialogue has been shown (the completion of the `verifyCard` function returns the `BTThreeDSecureCardNonce` card object which has the `liabilityShift...` check `BOOL` properties).
So you seem to contradict the documentation and indeed the actual SDK flow here. Could you clarify?

7. @demerino, you say above:
>Adding to what @sdcoffey said - you always need to run the threeDSecureVerification on a nonce from Drop-in and check the result to see if further verification is necessary.

I'm not sure what you mean by 'run the threeDSecureVerification on a nonce'. From what I can see, the only place anything called `threeDSecureVerification` exists within the SDK is as a property of the initial `BTDropInRequest` object used to initialise the `BTDropInController`, which we set to `true` or `false` simply to tell the Drop-In whether to do 3DS checks. What do you mean here by  'run the threeDSecureVerification'?

8. Ok, so this seems like a bit of a major flaw but maybe this is intentional. If the user selects a recent/saved card rather than adding a new card, 3DS is not shown at all. The `verifyCard` function is not even called. It seems to me that if 3DS is enabled, it should be for every transaction, regardless of whether the card is new or vaulted, especially since card vaulting is not necessarily unique to devices, and can be vaulted on one device and still shown on another. Thoughts?

9. Following on from this, in my earlier comment I said I thought it was odd to vault the card before 3DS if it was enabled. First off, is it possible to disable the vaulting at the drop-in when using a client token? I'm assuming not? Does it not feel like odd behaviour to you if your card details were saved, yet you had clicked cancel at 3DS and abandoned the transaction?
This behaviour actually appears to create a security hole when taking point 8 above into account. A user could start a transaction, get through to the 3DS dialogue, fail/cancel at this point, but then they are returned back to the start of the Drop-In where the vaulted card is now showing and is selectable. They can now select this as their payment method, and since 3DS checks are not performed on recent cards, their order will be processed successfully? Does this sound correct? Any thoughts?

Again, apologies for the absolute essay here but I'm just trying to be as clear as possible. Please let me know if you don't follow anything. Cheers! 👍

**Comment by sdcoffey** [comment]
Hey @meismyles,

This is awesome, we love the questions, and again, really appreciate your thoughtfulness here. Let me try to address these in the same order. 

1. You're right that an instance of `BTThreeDSecureDriver` is initialized by Drop-in, but you will need one of your own, for reasons I'll get to in the answering of a later question.

2. You are correct, if using Drop-in, you will not perform tokenization of `BTCard` objects directly. Drop-in will handle that for you, and return you the nonce as documented [here](https://developers.braintreepayments.com/guides/drop-in/ios/v4#starting-drop-in).

3. Great question. So the PR you submitted _did_ fix a bug in Drop-in. This section asking you implement `BTViewControllerPresentingDelegate` assumes you are not using Drop-in. When Drop-in is active, `BTCardFormViewController` would be the presenting controller, _not_ your top-level View  Controller. Basically, this means that if you were using Drop-in before we merged your PR, and did implement these delegate methods in your view controller, the ThreeDSecure web view controller still would not be shown, because as you can see [here](https://github.com/braintree/braintree-ios-drop-in/blob/master/BraintreeDropIn/BTCardFormViewController.m#L599), we set the presenting delegate to `this`, the current instance of `BTCardFormViewController`. However, if you're using 3DS in your app, you'll still need to implement this anyway, for reasons I'll get to in a later answer.

**Note:** I'd like to clarify here what it means to **verify** a card. In this context, verifying a card means that we're going to perform a lookup of a nonce created from a credit card, and determine if that card requires user authorization. If it does require user authorization, then the additional step of entering a password in the Cardinal web portal is required. For convenience, all of this functionality is handled for you in `[BTThreeDSecureDriver verifyCardWithNonce:]`.

4. So, to answer a couple of questions I've punted on here, if you're using 3DS, you _always_ need to verify the card in order to get the full benefits of 3DS. I think there was a little confusion here in a couple previous responses. Per your PR, Drop-in will verify the card at the time it is added, however, subsequent uses of that vaulted card need to be verified by you as well. See [this section in the docs](https://developers.braintreepayments.com/guides/3d-secure/client-side/ios/v4#verify-a-vaulted-credit-card), for more info. This is why, as per question 1, that you need to keep an instance of `BTThreeDSecureDriver` around even if you're using Drop-in, and that you'll also need to implement `BTViewControllerPresentingDelegate` in the View Controller that's responsible for presenting Drop-in.

5. It is true that we return a `BTPaymentMethodNonce *` to you in the callback from Drop-in, and the polymorphism is for a few reasons, the foremost being that Drop-in should be loosely coupled from specific payment method types, and that doing it this way lets us use the same callback generically for any payment method, which we think provides the nicest API for developers. Now, there are a couple parts to your question about how and whether you should check the type. If you're supporting more payment methods than just credit cards, you will need to do type checking and casting if you want more information about a payment method other than just the nonce. In the case of 3DS, as we said earlier, you'll always want to perform verification if you want to get the full benefits. This means that when you get a `BTPaymentMethodNonce` back from Drop-in, you'll want to check its type to ensure that it's a `BTCardNonce`, and pass its nonce to `[BTThreeDSecureDriver verifyCardWithNonce]`, which will call your `BTViewControllerPresentingDelegate` methods to present the ThreeDSecure verification view controller. 

6. So I think I misled you a little bit with my previous comment, sorry about that! So, to reiterate, if you're using 3DS, you _always_ need to perform verification via a `BTThreeDSecureDriver` instance when you have an instance of `BTCardNonce` that you want to transact against. This method will lookup the nonce, [determine if user authentication is required](https://github.com/braintree/braintree_ios/blob/master/Braintree3DSecure/BTThreeDSecureDriver.m#L88), and show the user the authentication view controller if so, as I alluded to in the note above answer 4.

7. I think this is more or less answered in the preceding answer and the note, but let me know if those were not clear.

8. This is indeed intentional. We intended Drop-in as a way for a user to _select_ a payment instrument, not to transact with one. To avoid feature creep/overreach in Drop-in, we require the merchant to perform additional verifications on vaulted cards if they elect to use 3DS.

9. There are a couple parts to this question as well. It's important to note that Client Tokens have different privilege levels, depending on whether you generated your Client Token with an associated [customer id](https://developers.braintreepayments.com/reference/request/client-token/generate/ruby#customer_id). If you did not use a customer id to generate your client token, Drop-in will _not_ vault your payment methods automatically. There's also a disconnect here between whether a card can be vaulted, and whether it can be transacted with. When we vault a card, all we do is store the card information in our database. We can then generate one or more nonces from that information, which you as the merchant use to create a transaction. Thus, 3DS verifications are tied to _transactions_, not backing cards. I realize that's a subtle distinction, but I hope that explains why we can and do vault a card regardless of whether 3DS verification is canceled or even succeeds. We perform verification in Drop-in when a card is added as a convenience for you so that you can use the nonce from a new card immediately after Drop-in returns.

**Comment by meismyles** [comment]
Hey @sdcoffey,

Thanks for your response. Many questions resolved, but more questions have arisen! 😂

So, based on your answers above, unless I am misunderstanding something (most likely), I can't help but think the Drop-in process behaves in an unexpected manner. Maybe there are use-cases involving other payment methods such as PayPal/Venmo/etc. that require its design be this way, but regardless, I'll elaborate and maybe you can resolve my confusion.

It is a bit difficult to convey easily in words, so I've made a pretty little diagram to hopefully keep us on the same page. The first flow shows what seems to me to be the current way the SDK works. The following 3 variations are methods that seem to me to be cleaner and/or more efficient and I wondered if you could explain the advantages the current implementation has over these other methods.
____
![braintree3dsvariations](https://cloud.githubusercontent.com/assets/6603912/23950265/54ad0d9a-0982-11e7-96e2-98b8d1c6c861.png)
___
**Current:** So in the current method, 3DS is called automatically by the Drop-in only if the user is adding a new card. You mention in point 4 of your previous comment:

>Per your PR, Drop-in will verify the card at the time it is added, however, subsequent uses of that vaulted card need to be verified by you as well.

So if the customer used a recent card, we now need to show them 3DS manually ourselves. This is fine, however, if the user _added a new card_, we will also be showing 3DS to them **again**. My first thought is that we could simply bypass 3DS if they added a new card, however, at the point where we enter the Drop-in completion block, unless I have missed something, it seems we have no way of knowing if the customer added a new card or tapped a recent card. What am I missing here?

<br>

**Variation 1:** This variation is essentially exactly the solution I describe in the paragraph above. A minor change to the SDK that means we know in the Drop-in completion block whether the customer added a new card or used a recent card. This then determines whether or not we manually show 3DS to the customer.

Since it is a minor change, my colleague has submitted a [pull request](https://github.com/braintree/braintree-ios-drop-in/pull/29) simply to illustrate better what this would look like.

_Note: Regardless of the use/validity of this pull request in terms of this 3DS discussion, it actually offers another separate benefit. I had previously emailed your support team wondering if it was possible in the Drop-in completion to determine whether a new card or recent card had been used by the customer, so that we could use this information for our analytics (to see how many people were using recent cards). They said it was currently not possible to do this from within the app, but I guess this raises a good opportunity to hear your thoughts on it and potentially add this functionality in. See issue #28._

<br>

**Variation 2:** So this is the variation that I originally thought was how the SDK was working. If the developer enables `threeDSecureVerification` in the `BTDropInRequest`, then they need to do nothing other than set the `request` amount. This still seems to me to be the cleanest solution. Enabled - 3DS shows for both new and recent cards. Disabled - 3DS doesn't show.

If you already show 3DS automatically when adding new cards, why not also just show it automatically for recent cards as well to avoid confusion? This variation seems like it would require the biggest changes, but personally it seems like it is the cleanest for prospective developers looking to implement Braintree with 3DS.

<br>

**Variation 3:** This variation is not really a method that concerns you guys as developers of the Braintree SDK, but rather what I believe is a cleaner solution for us if the SDK remains the way it is. As mentioned above, it seems that we definitely need to do 3DS checks manually in the Drop-in completion. As such, it seems cleaner to me to just disable `threeDSecureVerification` in the `BTDropInRequest` so that 3DS **doesn't** show automatically at the end of adding a new card, and then let the manual 3DS checks we are doing in the completion just handle this for both new and recent cards. Since we will already be implementing `BTViewControllerPresentingDelegate` in our own view controller, 3DS can just show over this for both new and recent. Any thoughts?

**Comment by sdcoffey** [comment]
@meismyles,

Take a look [this PR](https://github.com/braintree/braintree-ios-drop-in/pull/30) - @demerino and I spent a while today looking over our 3DS integration with respect to Drop-in, and we're considering this small change. When a user cancels a 3DS flow inside of Drop-in, when adding a new card, `BTDropInResult.canceled` will be set to `true`.

The Drop-in demo now includes an example for 3DS which checks the nonce type and starts the verification flow if necessary. This obviously doesn't solve all the questions about how Drop-in integrates with 3DS, but we hope that it's a start. We're also going to consider larger changes to the 3DS+Drop-in integration based on some of the feedback you've laid out.

**Comment by sdcoffey** [comment]
@meismyles 

If you're still having issues related to 3DS, or anything else, please contact me directly. My email is steve.coffey@getbraintree.com. It will probably be quicker to answer your questions over Google Hangouts than to go back and forth on GH.

**Comment by demerino** [comment]
To simulate the 3DS integration in the `demo` - you'll need to adjust the settings to use a client token, customer and require 3DS. See the example below.

<img width="312" alt="3ds-example-ss" src="https://cloud.githubusercontent.com/assets/7817434/23976186/56c3952e-09a3-11e7-956b-10dc924ed255.png">

**Comment by meismyles** [comment]
@sdcoffey Ok cheers, will message you on hangouts.

**Comment by kierobinson1** [comment]
@sdcoffey @demerino  We've just come across another issue with the `BTThreeDSecureAuthenticationViewControllerDelegate` functions that are called on completion of the 3DS web view:
```objc
BTThreeDSecureDriver.m - line 175

- (void)threeDSecureViewController:(__unused BTThreeDSecureAuthenticationViewController *)viewController
               didAuthenticateCard:(BTThreeDSecureCardNonce *)tokenizedCard
                        completion:(void (^)(BTThreeDSecureViewControllerCompletionStatus))completionBlock
{
    self.upgradedTokenizedCard = tokenizedCard;
    completionBlock(BTThreeDSecureViewControllerCompletionStatusSuccess);
    [self.apiClient sendAnalyticsEvent:@"ios.threedsecure.authenticated"];
}

- (void)threeDSecureViewController:(__unused BTThreeDSecureAuthenticationViewController *)viewController
                  didFailWithError:(NSError *)error {
    if ([error.domain isEqualToString:BTThreeDSecureErrorDomain] && error.code == BTThreeDSecureErrorTypeFailedAuthentication) {
        [self.apiClient sendAnalyticsEvent:@"ios.threedsecure.error.auth-failure"];
    } else {
        [self.apiClient sendAnalyticsEvent:@"ios.threedsecure.error.unrecognized-error"];
    }
    self.upgradedTokenizedCard = nil;
    self.completionBlockAfterAuthenticating(nil, error);
}

- (void)threeDSecureViewControllerDidFinish:(BTThreeDSecureAuthenticationViewController *)viewController {
    if (self.upgradedTokenizedCard) {
        self.completionBlockAfterAuthenticating(self.upgradedTokenizedCard, nil);
    } else {
        self.completionBlockAfterAuthenticating(nil, nil);
        [self.apiClient sendAnalyticsEvent:@"ios.threedsecure.canceled"];
    }
    self.completionBlockAfterAuthenticating = nil;
    [self informDelegateRequestsDismissalOfViewController:viewController];
}
```
`didAuthenticateCard` doesn't call `self.completionBlockAfterAuthenticating` so when `threeDSecureViewControllerDidFinish` does call `self.completionBlockAfterAuthenticating`, it is only called **once** which seems correct.

However, `didFailWithError` calls `self.completionBlockAfterAuthenticating` as well as `threeDSecureViewControllerDidFinish` calling it, which means if there is an error, the completion block gets called **twice**. The first time an error is passed back. The second time, `threeDSecureViewControllerDidFinish` assumes the user cancelled and so passes back `(nil, nil)`.

**Comment by meismyles** [comment]
Test card from Cardinal used to recreate this:
* **PAN:** 4000 0000 0000 0028
* **Expiry:** 01/20
* **CVV:** 111

**Comment by demerino** [comment]
`Braintree 4.8.0` should solve the double callback bug.

**Comment by demerino** [comment]
Fixed in `5.2.0`.
